### ISSUE TITLE
There is some discrepancy between the document explain and code implement of optim.SGD() when use maximize=True

### ISSUE THREAD

**Issue by l1351868270** [pytorch/pytorch#149476]
### 🐛 Describe the bug

In the document (https://pytorch.org/docs/stable/generated/torch.optim.SGD.html) .The loop is 

![Image](https://github.com/user-attachments/assets/e9f7e320-dcfc-4911-b7d1-d94444b76849)

It means:
1.get the original grad,  $g_{t} = og_{t}$
2.weight_decay  != 0, update the grad, $g_{t} = og_{t} + λθ_{t-1}$
3.momentum != 0, update the grad, 
$b_{t} = μb_{t-1} + (1-τ)g_{t} = μb_{t-1} + (1-τ)(og_{t} + λθ_{t-1})$ , 
then 
$g_{t} = b_{t} = μb_{t-1} + (1-τ)(og_{t} + λθ_{t-1})$
or  
$g_{t} = g_{t} + μb_{t} = g_{t} + μ(μb_{t-1} + (1-τ)g_{t}) = μ^{2}b_{t-1} + (1 + μ(1-τ))g_{t} =  μ^{2}b_{t-1} + (1 + μ(1-τ))(og_{t} + λθ_{t-1})$
4.if maximize == True, update the params, 
$θ_{t} = θ_{t-1} + γ(μb_{t-1} + (1-τ)(og_{t} + λθ_{t-1})) = (1 + γ(1-τ)λ)θ_{t-1} + γμb_{t-1} + γ(1-τ)og_{t}$
or
$θ_{t} = θ_{t-1} + γ(μ^{2}b_{t-1} + (1 + μ(1-τ))(og_{t} + λθ_{t-1})) =(1 + γ(1 + μ(1-τ))λ)θ_{t-1}  + γμ^{2}b_{t-1} +  γ(1 + μ(1-τ))og_{t}$

In the code implement,The loop is 

```
  for i, param in enumerate(params):
        grad = grads[i] if not maximize else -grads[i]

        if weight_decay != 0:
            # Nested if is necessary to bypass jitscript rules
            if isinstance(weight_decay, Tensor):
                if weight_decay.requires_grad:
                    # usually this is the differentiable path, which is why the param.clone() is needed
                    grad = grad.addcmul_(param.clone(), weight_decay)
                else:
                    grad = grad.add(param, alpha=weight_decay)
            else:
                grad = grad.add(param, alpha=weight_decay)

        if momentum != 0:
            buf = momentum_buffer_list[i]

            if buf is None:
                buf = torch.clone(grad).detach()
                momentum_buffer_list[i] = buf
            else:
                buf.mul_(momentum).add_(grad, alpha=1 - dampening)

            if nesterov:
                grad = grad.add(buf, alpha=momentum)
            else:
                grad = buf

        # Nested if is necessary to bypass jitscript rules
        if isinstance(lr, Tensor):
            if lr.requires_grad:
                param.addcmul_(grad, lr, value=-1)
            else:
                param.add_(grad, alpha=-lr)
        else:
            param.add_(grad, alpha=-lr)
```

It means:
1.get the original grad,  $g_{t} = -og_{t}$
2.weight_decay  != 0, update the grad, $g_{t} = -og_{t} + λθ_{t-1}$
3.momentum != 0, update the grad, 
$b_{t} = μb_{t-1} + (1-τ)g_{t} = μb_{t-1} + (1-τ)(-og_{t} + λθ_{t-1})$ , 
then 
$g_{t} = b_{t} = μb_{t-1} + (1-τ)(-og_{t} + λθ_{t-1})$
or  
$g_{t} = g_{t} + μb_{t} = g_{t} + μ(μb_{t-1} + (1-τ)g_{t}) = μ^{2}b_{t-1} + (1 + μ (1-τ))g_{t} = μ^{2}b_{t-1} + (1 + μ (1-τ))(-og_{t} + λθ_{t-1})$
4.if maximize == True, update the params, 
$θ_{t} = θ_{t-1} - γ(μb_{t-1} + (1-τ)(-og_{t} + λθ_{t-1})) = (1-γ(1-τ)λ)θ_{t-1} - γμb_{t-1} + γ(1-τ)og_{t}$
or
$θ_{t} = θ_{t-1} - γ(μ^{2}b_{t-1} + (1 + μ (1-τ))(-og_{t} + λθ_{t-1}))= (1-γ(1 + μ (1-τ))λ)θ_{t-1}- γμ^{2}b_{t-1} + γ(1 + μ (1-τ))og_{t}$

we could find that, only the $og_{t}$ item is the same. the $θ_{t-1}$ item and $b_{t-1}$ item are all diffrent.

i also implement one based on the document(https://github.com/l1351868270/ld_triton/blob/main/ld_triton/optim/sgd/naive_sgd.py), when maximize == True, the result is diffrent from the pytorch implement.

### Versions

master

cc @svekars @sekyondaMeta @AlannaBurke @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

**Comment by janeyx99** [comment]
We'd accept a PR to align the documentation to what the code does

**Comment by l1351868270** [comment]
@janeyx99 look for these two formula carfully
$θ_{t} = (1 + γ(1-τ)λ)θ_{t-1} + γμb_{t-1} + γ(1-τ)og_{t}$ 
$θ_{t} = (1-γ(1-τ)λ)θ_{t-1} - γμb_{t-1} + γ(1-τ)og_{t}$

when $λ=0$
$θ_{t} =γμb_{t-1} + γ(1-τ)og_{t}$ 
$θ_{t} = - γμb_{t-1} + γ(1-τ)og_{t}$
these two formula is equal, because of the document is $b_{t-1} = linear(og)$ and the code implement is $b_{t-1} = linear(-og)$
via
$θ_{t} = linear(og) + γ(1-τ)og_{t}$ 
$θ_{t} =  linear(og) + γ(1-τ)og_{t}$

the only difference is the $θ_{t-1}$ item when $λ \neq 0$

the weight decay theroy is that minimize the weight 

when maximize  is false
$L = L_{0} + \frac{λ}{2} \|θ\|^2$ then $\frac{\partial L}{\partial θ } = \frac{\partial L_{0}}{\partial θ } + λθ$
and when maximize  is true
$L = L_{0} - \frac{λ}{2} \|θ\|^2$ then $\frac{\partial L}{\partial θ } = \frac{\partial L_{0}}{\partial θ } - λθ$

so, the current document only change the $g_{t} = g_{t} - λθ_{t-1}$ when maximize  is true, it will equal the implement code

i also change the implement (https://github.com/l1351868270/ld_triton/blob/main/ld_triton/optim/sgd/naive_sgd.py) when maximize == True, the result will same as the pytorch implement.
```
        if weight_decay != 0:
            if maximize:
                grad = grad.add(param, alpha=-weight_decay)
            else:
                grad = grad.add(param, alpha=weight_decay)
```
